common/utf8validator/validate.go

```diff
@@ -0,0 +1,250 @@
+	checkPersistence dynamicconfig.FloatPropertyFn,
+	failRPCRequest dynamicconfig.BoolPropertyFn,
+	failRPCResponse dynamicconfig.BoolPropertyFn,
+	failPersistence dynamicconfig.BoolPropertyFn,
+) *Validator {
+	v := &Validator{
+		logger:  logger,
+		metrics: metrics,
+		checkBySource: [sourceCount]dynamicconfig.FloatPropertyFn{
+			checkRPCRequest, checkRPCResponse, checkPersistence,
+		},
+		failBySource: [sourceCount]dynamicconfig.BoolPropertyFn{
+			failRPCRequest, failRPCResponse, failPersistence,
+		},
+	}
+	globalValidator.Store(v)
+	return v
+}
+
+// ValidateUsingGlobalValidator maybe validates one proto message, depending on dynamic config
+// and random sampling. If validation fails, this may still return no error, depending on
+// dynamic config. This deliberately does not return a serviceerror because the decision of
+// which serviceerror to use is contextual. The caller should wrap it in the appropriate
+// serviceerror.
+func ValidateUsingGlobalValidator(
+	m proto.Message,
+	source MessageSource,
+	getTags func() []tag.Tag,
+) error {
+	if v, ok := globalValidator.Load().(*Validator); ok {
+		return v.Validate(m, source, getTags)
+	}
+	return nil
+}
+
+// Validate maybe validates one proto message, depending on dynamic config and random sampling.
+// If validation fails, this may still return no error, depending on dynamic config. This
+// deliberately does not return a serviceerror because the decision of which serviceerror to
+// use is contextual. The caller should wrap it in the appropriate serviceerror.
+func (v *Validator) Validate(
+	m proto.Message,
+	source MessageSource,
+	getTags func() []tag.Tag,
+) error {
+	if source < 0 || source >= sourceCount {
+		return ErrInvalidMessageSource
+	}
+	checkFraction := v.checkBySource[source]()
+	// This makes checkFraction = 0.0 really mean "never" and checkFraction = 1.0 really mean "always".
+	cutoff := uint64(checkFraction * (1 << 32))
+	if rand.Uint64()&(1<<32-1) >= cutoff {
+		return nil
+	}
+
+	var validation validation
+	ref := m.ProtoReflect()
+	if validation.validateMessage(ref) {
+		return nil
+	}
+
+	badPath := string(ref.Descriptor().Name()) + "." + validation.badPath
+
+	var tags []tag.Tag
+	if getTags != nil {
+		tags = getTags()
+	}
+	v.logger.Warn("invalid utf-8 in string field", append(tags,
+		tag.NewStringTag("message-source", sourceNames[source]),
+		tag.NewStringTag("message-path", badPath),
+	)...)
+	metrics.UTF8ValidationErrors.With(v.metrics).Record(1)
+
+	if v.failBySource[source]() {
+		return ErrInvalidUTF8
+	}
+	return nil
+}
+
+var _ grpc.UnaryServerInterceptor = (*Validator)(nil).Intercept
+
+// Intercept acts as grpc interceptor.
+func (v *Validator) Intercept(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	getTags := func() []tag.Tag {
+		if getNs, ok := req.(interface{ GetNamespace() string }); ok {
+			return []tag.Tag{
+				tag.WorkflowNamespace(getNs.GetNamespace()),
```

Comment:
I had it that way first, but switched to the lazy approach since we only need it if something fails, which we don't expect to ever happen, so it's fine to put more work on that path. It's not expensive (these are proto messages, it's just an accessor), but I wanted to make the fast path as fast as possible.

Reply:
@gow `GetNamespace` is extremely cheap as it's just a getter method generated by the golang protobuf plugin. They're around four lines long:

```
type Frobnicator struct {
    Namespace string
}

func (f *Frobnicator) GetNamespace() string {
    if f == nil {
        return ""
    }
    return f.Namespace
}
```